### PR TITLE
LP-2458 Unit tests for update_is_prerequisite_courses_passed command added

### PR DESCRIPTION
**Pull request by annas-usman**
**test_update_is_prerequisite_courses_passed.py**. added which contains tests for management command

Here is the [Ticket LP-2458](https://philanthropyu.atlassian.net/browse/LP-2458)

### CODE REVIEW

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/557/workflows/workflows_id/jobs/557/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/common/course_meta/factory.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_helpers.py (100%)
openedx/adg/lms/applications/__init__.py (100%)
openedx/adg/lms/applications/admin.py (100%)
openedx/adg/lms/applications/apps.py (100%)
openedx/adg/lms/applications/constants.py (100%)
openedx/adg/lms/applications/handlers.py (100%)
openedx/adg/lms/applications/helpers.py (100%)
openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py (100%)
openedx/adg/lms/applications/models.py (95.8%): Missing lines 54
openedx/adg/lms/applications/test/factories.py (100%)
openedx/adg/lms/applications/test/test_handlers.py (100%)
openedx/adg/lms/applications/test/test_helpers.py (100%)
openedx/adg/lms/registration_extension/forms.py (100%)
openedx/adg/lms/registration_extension/models.py (100%)
openedx/adg/lms/user_authn_override/login_form.py (100%)
-------------
Total:   186 lines
Missing: 1 line
Coverage: 99%
-------------
```

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/563/workflows/workflows_id/jobs/563/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/common/course_meta/factory.py (100%)
openedx/adg/common/mailchimp_pipeline/tests/test_helpers.py (100%)
openedx/adg/lms/applications/__init__.py (100%)
openedx/adg/lms/applications/admin.py (100%)
openedx/adg/lms/applications/apps.py (100%)
openedx/adg/lms/applications/constants.py (100%)
openedx/adg/lms/applications/handlers.py (100%)
openedx/adg/lms/applications/helpers.py (100%)
openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py (100%)
openedx/adg/lms/applications/models.py (95.8%): Missing lines 54
openedx/adg/lms/applications/test/factories.py (100%)
openedx/adg/lms/applications/test/test_handlers.py (100%)
openedx/adg/lms/applications/test/test_helpers.py (100%)
openedx/adg/lms/registration_extension/forms.py (100%)
openedx/adg/lms/registration_extension/models.py (100%)
openedx/adg/lms/user_authn_override/login_form.py (100%)
-------------
Total:   186 lines
Missing: 1 line
Coverage: 99%
-------------
```

File: openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+    assert wrapped_exception.type == SystemExit
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize('enroll_user',
+                         [dict({'is_active': False})],
+                         indirect=True)
+def test_handle_with_no_minimal_user(user_for_testing, enroll_user):
+
```
Comment (zahid-arbisoft) [inline openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py:134]: Lets remove this blank line.

File: openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        is_active=request.param['is_active']
+    )
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize('enroll_user',
+                         [dict({'is_active': False})],
+                         indirect=True)
+def test_minimal_users_with_inactive_user(enroll_user):
+    pre_req_courses = CourseMeta.open_pre_req_courses.all()
```
Comment (zahid-arbisoft) [inline openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py:52]: Please add a short docstring for each unit test. For unit test docstring we do not add params or return type in docstring.

```
"""
My test docstring
"""
```

File: openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+    assert user_application_hub.is_prerequisite_courses_passed == False
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize('enroll_user',
+                         [dict({'is_active': True})],
+                         indirect=True)
+def test_handle_with_user_failed(user_for_testing, enroll_user):
+
```
Comment (zahid-arbisoft) [inline openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py:148]: Same, lets remove the blank line from here as well

**Review by zahid-arbisoft — CHANGES_REQUESTED** [review]

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/564/workflows/workflows_id/jobs/564/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/common/mailchimp_pipeline/tests/test_helpers.py (100%)
openedx/adg/lms/applications/__init__.py (100%)
openedx/adg/lms/applications/admin.py (100%)
openedx/adg/lms/applications/apps.py (100%)
openedx/adg/lms/applications/constants.py (100%)
openedx/adg/lms/applications/handlers.py (100%)
openedx/adg/lms/applications/helpers.py (100%)
openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py (100%)
openedx/adg/lms/applications/models.py (95.8%): Missing lines 54
openedx/adg/lms/applications/test/factories.py (100%)
openedx/adg/lms/applications/test/test_handlers.py (100%)
openedx/adg/lms/applications/test/test_helpers.py (100%)
openedx/adg/lms/registration_extension/forms.py (100%)
openedx/adg/lms/registration_extension/models.py (100%)
openedx/adg/lms/user_authn_override/login_form.py (100%)
-------------
Total:   179 lines
Missing: 1 line
Coverage: 99%
-------------
```

File: openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    pre_req_courses = CourseMeta.open_pre_req_courses.all()
+
+    command = command_module.Command()
+    minimal_users = command.get_minimal_users_to_be_checked_for_update(pre_req_courses)
+
+    assert minimal_users.count() == 0
+
+
+@pytest.mark.django_db
+def test_minimal_users_none_registered_in_prerequisite(course_for_testing):
```
Comment (zahid-arbisoft) [inline openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py:61]: Please rename test like this "test_{name_of_function}_{test_case}"

`test_get_minimal_users_to_be_checked_for_update_not_registered_in_prerequisite`

File: openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+def test_minimal_users(enroll_user):
+    """
+    Test to check that an active user registered in prerequisite is counted while calculating minimal users
+    """
+    pre_req_courses = CourseMeta.open_pre_req_courses.all()
+
+    command = command_module.Command()
+    minimal_users = command.get_minimal_users_to_be_checked_for_update(pre_req_courses)
+
+    assert minimal_users.count() > 0
```
Comment (zahid-arbisoft) [inline openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py:84]: Please add test cases for the following scenarios:

Create two users, user 1 and user 2, leave user 1 as-is and wherever required, enroll user 2 to course(s).

Create three courses of them marked two pre-req

1. Enroll in two courses, enrollment active in both, and pass both courses as param (list) `pre_req_courses` to `get_minimal_users_to_be_checked_for_update`. Test should user 2 

2. Enroll in three courses, enrollment active in two, and pass both courses as param (list) `pre_req_courses` to `get_minimal_users_to_be_checked_for_update`. Test should user 2 

3. Enroll in two courses, enrollment active in two, and pass two courses (one in which you enrolled and one in which you did not enroll) as param (list) `pre_req_courses` to `get_minimal_users_to_be_checked_for_update`. Test should not return any user

4. Enroll in two courses, enrollment active in two, and pass two courses (one in which you enrolled and one in which you did not enroll) as param (list) `pre_req_courses` to `get_minimal_users_to_be_checked_for_update`. Test should not return any user

5. Enroll in two courses, enrollment active in both, and pass all three courses as param (list) `pre_req_courses` to `get_minimal_users_to_be_checked_for_update`. Test should not return any user

6. In application hub set `is_prerequisite_courses_passed=true`, Enroll in two courses, enrollment active in both, and pass both courses as param (list) `pre_req_courses` to `get_minimal_users_to_be_checked_for_update`. Test should not return any user.

File: openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize('enroll_user', [dict({'is_active': True})], indirect=True)
+def test_user_failed_in_course(user_for_testing, enroll_user):
+    """
+    Test to check if the helper function returns True for a user not passed in course
+    """
+    pre_req_courses = CourseMeta.open_pre_req_courses.all()
```
Comment (zahid-arbisoft) [inline openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py:93]: This is not required in this test `pre_req_courses = CourseMeta.open_pre_req_courses.all()`

File: openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+    command = command_module.Command()
+    minimal_users = command.get_minimal_users_to_be_checked_for_update(pre_req_courses)
+
+    assert minimal_users.count() > 0
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize('enroll_user', [dict({'is_active': True})], indirect=True)
```
Comment (zahid-arbisoft) [inline openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py:88]: Not required `@pytest.mark.parametrize ...`

Though you can change it to param True for passed and CourseGradeFactory read returning None

File: openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+def test_user_failed_in_course(user_for_testing, enroll_user):
+    """
+    Test to check if the helper function returns True for a user not passed in course
+    """
+    pre_req_courses = CourseMeta.open_pre_req_courses.all()
+
+    command = command_module.Command()
+    is_user_failed = command.is_user_failed_in_course(user_for_testing, pre_req_courses.first())
+
+    assert is_user_failed == True
```
Comment (zahid-arbisoft) [inline openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py:98]: 1st test case:
Please mock `CourseGradeFactory().read` once you do that, you do not need params, you can pass `mocker.ANY`

```
mocked_course_grade_factory_read = ...
mocked_course_grade_factory_read.return_value = True

command.is_user_failed_in_course(mocker.ANY, mock.ANY)
```

2nd test case:

There will be another test case, to test function returns false if `course_grade` is None

```
mocked_course_grade_factory_read = ...
mocked_course_grade_factory_read.return_value = None

command.is_user_failed_in_course(mocker.ANY, mock.ANY)
```

File: openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize('enroll_user', [dict({'is_active': True})], indirect=True)
+def test_user_passed_in_course(mocker, user_for_testing, enroll_user):
+    """
+    Test to check if the helper function returns False for a user passed in course
```
Comment (zahid-arbisoft) [inline openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py:105]: No need for this test case

File: openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    pre_req_courses = CourseMeta.open_pre_req_courses.all()
+
+    command = command_module.Command()
+    is_user_failed = command.is_user_failed_in_course(user_for_testing, pre_req_courses.first())
+
+    assert is_user_failed == False
+
+
+@pytest.mark.django_db
+def test_handle_with_no_prerequisites():
```
Comment (zahid-arbisoft) [inline openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py:123]: `test_command_update_is_prerequisite_courses_passed_with_no_prerequisites`

Please update other core command related test names on similar pattern

File: openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+@pytest.mark.parametrize('enroll_user', [dict({'is_active': False})], indirect=True)
+def test_handle_with_no_minimal_user(user_for_testing, enroll_user):
+    """
+    Test to check if management command does not update `is_prerequisite_courses_passed` flag if minimal users are zero
+    """
+    call_command('update_is_prerequisite_courses_passed')
+
+    user_application_hub, _ = ApplicationHub.objects.get_or_create(user=user_for_testing)
+
+    assert user_application_hub.is_prerequisite_courses_passed == False
```
Comment (zahid-arbisoft) [inline openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py:143]: Please mock `get_minimal_users_to_be_checked_for_update` and `is_user_failed_in_course` while writing test for core command, since these two methods are tested spertately

**Review by zahid-arbisoft — CHANGES_REQUESTED** [review]

**Review by ahmed-faraz46 — APPROVED** [review]
Brilliant identification of test cases and implementation 👏 Here are a few minor suggestions.

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/615/workflows/workflows_id/jobs/615/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/common/mailchimp_pipeline/tests/test_helpers.py (100%)
openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py (100%)
-------------
Total:   140 lines
Missing: 0 lines
Coverage: 100%
-------------
```

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/616/workflows/workflows_id/jobs/616/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py (100%)
-------------
Total:   138 lines
Missing: 0 lines
Coverage: 100%
-------------
```

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/617/workflows/workflows_id/jobs/617/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py (100%)
-------------
Total:   132 lines
Missing: 0 lines
Coverage: 100%
-------------
```

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/618/workflows/workflows_id/jobs/618/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py (100%)
-------------
Total:   132 lines
Missing: 0 lines
Coverage: 100%
-------------
```

File: openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+    user = create_users.pop()
+    enrollments = list()
+
+    for i, is_active in enumerate(activations):
+        enrollment = CourseEnrollmentFactory(
+            course_id=courses[i].id,
+            user=user,
+            is_active=is_active
+        )
+        enrollments.append(enrollment)
```
Comment (zahid-arbisoft) [inline openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py:67]: Do we need `enrollments` list since we are not using it anywhere

File: openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+    return prerequisites, user
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize('create_users', [1], indirect=True)
+@pytest.mark.parametrize('create_courses', [[True]], indirect=True)
+@pytest.mark.parametrize('enroll_users, expected_result', [([False], 0), ([True], 1)], indirect=['enroll_users'],
+                         ids=['enrolled', 'not_enrolled'])
+def test_get_minimal_users_to_be_checked_for_update_for_active_and_inactive_enrollments(enroll_users, expected_result):
+    # pylint: disable=redefined-outer-name
```
Comment (MunamMubashir-arbisoft) [inline openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py:78]: Instead of disabling these errors, solve them by giving a `name` to your fixtures.
```
@pytest.fixture(name='test_users')
def enroll_users(request, create_users, create_courses):
```

And then pass whatever name you've given it, to the tests where that particular fixture is needed.

File: openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+@pytest.mark.parametrize('enroll_users', [[True, True]], indirect=True)
+def test_get_minimal_users_to_be_checked_for_update_case_with_prerequisites_already_passed(enroll_users):
+    # pylint: disable=redefined-outer-name
+    """
+    Test to check that a user already passed in prerequisites and a user with no enrollments are not counted while
+    calculating minimal users
+    """
+    prerequisites, user = enroll_users
+
+    user_application_hub, _ = ApplicationHub.objects.get_or_create(user=user)
```
Comment (zahid-arbisoft) [inline openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py:192]: Lets create a factory for application hub using factory boy and use [get_or_create](https://factoryboy.readthedocs.io/en/2.12.0/orms.html#factory.django.DjangoOptions.django_get_or_create)

**Review by zahid-arbisoft — CHANGES_REQUESTED** [review]

File: openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+    minimal_users = command.get_minimal_users_to_be_checked_for_update(prerequisites)
+
+    assert minimal_users.count() == 0
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize('create_users', [2], indirect=True)
+@pytest.mark.parametrize('create_courses', [[True, True]], indirect=True)
+@pytest.mark.parametrize('enroll_users', [[True, True]], indirect=True)
+def test_get_minimal_users_to_be_checked_for_update_with_2_active_enrollments(enroll_users):
```
Comment (zahid-arbisoft) [inline openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py:110]: Please merge tests where possible

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/625/workflows/workflows_id/jobs/625/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py (100%)
openedx/adg/lms/applications/test/factories.py (100%)
-------------
Total:   106 lines
Missing: 0 lines
Coverage: 100%
-------------
```

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/626/workflows/workflows_id/jobs/626/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py (100%)
openedx/adg/lms/applications/test/factories.py (100%)
-------------
Total:   106 lines
Missing: 0 lines
Coverage: 100%
-------------
```

**Review by zahid-arbisoft — APPROVED** [review]

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/629/workflows/workflows_id/jobs/629/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/anas_usman/LP-2458_dev...HEAD, staged and unstaged changes
-------------
openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py (100%)
openedx/adg/lms/applications/test/factories.py (100%)
-------------
Total:   105 lines
Missing: 0 lines
Coverage: 100%
-------------
```

**Comment by zahid-arbisoft** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/630/workflows/workflows_id/jobs/630/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/lms/applications/management/commands/tests/test_update_is_prerequisite_courses_passed.py (100%)
openedx/adg/lms/applications/test/factories.py (100%)
-------------
Total:   105 lines
Missing: 0 lines
Coverage: 100%
-------------
```

**Review by ahmed-faraz46 — APPROVED** [review]
